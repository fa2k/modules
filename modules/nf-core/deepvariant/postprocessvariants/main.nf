process DEEPVARIANT_POSTPROCESSVARIANTS {
    tag "$meta.id"
    label 'process_single'

    //Conda is not supported at the moment
    container "nf-core/deepvariant:1.5.0"

    input:
    tuple val(make_examples_id), val(meta), path(variant_calls_tfrecord), path(gvcf_tfrecords)
    tuple val(meta2), path(fasta)
    tuple val(meta3), path(fai)
    tuple val(meta4), path(gzi)

    output:
    tuple val(meta), path("${prefix}.vcf.gz")      ,  emit: vcf
    tuple val(meta), path("${prefix}.vcf.gz.tbi")  ,  emit: vcf_tbi
    tuple val(meta), path("${prefix}.g.vcf.gz")    ,  emit: gvcf
    tuple val(meta), path("${prefix}.g.vcf.gz.tbi"),  emit: gvcf_tbi

    path "versions.yml",  emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    // Exit if running this module with -profile conda / -profile mamba
    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
        error "DEEPVARIANT module does not support Conda. Please use Docker / Singularity / Podman instead."
    }
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"

    def matcher = gvcf_tfrecords[0].baseName =~ /^(.+)-\d{5}-of-(\d{5})$/
    if (!matcher.matches()) {
        throw new IllegalArgumentException("tfrecord baseName '" + gvcf_tfrecords[0].baseName + "' doesn't match the expected pattern")
    }
    def gvcf_tfrecord_name = matcher[0][1]
    def shardCount = matcher[0][2]
    // Reconstruct the logical name - ${tfrecord_name}.examples.tfrecord@${task.cpus}.gz
    def gvcf_tfrecords_logical_name = "${gvcf_tfrecord_name}@${shardCount}.gz"

    """
    /opt/deepvariant/bin/postprocess_variants \\
        --ref=${fasta} \\
        --infile=${variant_calls_tfrecord} \\
        --outfile=${prefix}.vcf.gz \\
        --nonvariant_site_tfrecord_path ${gvcf_tfrecords_logical_name} \\
        --gvcf_outfile=${prefix}.g.vcf.gz \\
        ${args}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        deepvariant_postprocessvariants: \$(echo \$(/opt/deepvariant/bin/run_deepvariant --version) | sed 's/^.*version //; s/ .*\$//' )
    END_VERSIONS
    """

    stub:
    // Exit if running this module with -profile conda / -profile mamba
    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
        error "DEEPVARIANT module does not support Conda. Please use Docker / Singularity / Podman instead."
    }
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    echo "" | gzip > ${prefix}.vcf.gz
    touch ${prefix}.vcf.gz.tbi
    echo "" | gzip > ${prefix}.g.vcf.gz
    touch ${prefix}.g.vcf.gz.tbi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        deepvariant_callvariants: \$(echo \$(/opt/deepvariant/bin/run_deepvariant --version) | sed 's/^.*version //; s/ .*\$//' )
    END_VERSIONS
    """
}
