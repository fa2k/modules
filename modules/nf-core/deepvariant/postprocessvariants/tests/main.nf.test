nextflow_process {

    name "Test Process DEEPVARIANT_POSTPROCESSVARIANTS"
    script "../main.nf"
    process "DEEPVARIANT_POSTPROCESSVARIANTS"

    tag "deepvariant/postprocessvariants"
    tag "deepvariant"
    tag "modules"
    tag "modules_nfcore"

    test("homo_sapiens - wgs") {
        setup {
            run("DEEPVARIANT_MAKEEXAMPLES") {
                script "../../makeexamples/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test', single_end:false ], // meta map
                        file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                        file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true),
                        []
                    ]
                    input[1] = [
                        [ id:'genome'],
                        file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                    ]
                    input[2] = [
                        [ id:'genome'],
                        file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                    ]
                    input[3] = [
                        [],[]
                    ]
                    """
                }
            }
            run("DEEPVARIANT_CALLVARIANTS") {
                script "../../callvariants/main.nf"
                process {
                    """
                    input[0] = DEEPVARIANT_MAKEEXAMPLES.out.examples
                    input[1] = [
                        "wgs"
                    ]
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = DEEPVARIANT_CALLVARIANTS.out.call_variants_tfrecords.join(
                    DEEPVARIANT_MAKEEXAMPLES.out.gvcf,
                    by: [0, 1],
                    failOnMismatch: true
                )
                input[1] = [
                    [ id:'genome'],
                    file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ]
                input[2] = [
                    [ id:'genome'],
                    file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                ]
                input[3] = [
                    [],[]
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("homo_sapiens - wgs - stub") {
        options "-stub"
        
        when {
            process {
                """
                input[0] = [
                    'make_examples_id',
                    [ id:'test', single_end:false ],
                    [],
                    "gvcf.tfrecord@32.gz",
                    [],
                ]
                input[1] = [
                    [ id:'genome'],
                    file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ]
                input[2] = [
                    [ id:'genome'],
                    file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                ]
                input[3] = [
                    [],[]
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match("stub") }
            )
        }
    }

}