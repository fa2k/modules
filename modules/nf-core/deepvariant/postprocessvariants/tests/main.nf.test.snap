{
    "stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.g.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.g.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,72b3f70b4108ecc2a8744940b5b513d7"
                ],
                "gvcf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.g.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "gvcf_tbi": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.g.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,72b3f70b4108ecc2a8744940b5b513d7"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T16:58:08.641342"
    },
    "homo_sapiens - wgs": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.vcf.gz:md5,7cba1516f7cf0888765d5ee8caf275f4"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.vcf.gz.tbi:md5,02a78562bc83520a51010a01fb06f217"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.g.vcf.gz:md5,8d6ac08997bfd848a0a4d9d295e76952"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.g.vcf.gz.tbi:md5,37e2d8f4cca0a21113cede608f54885a"
                    ]
                ],
                "4": [
                    "versions.yml:md5,27500aacf5e1da6a8a31905566fcb41f"
                ],
                "gvcf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.g.vcf.gz:md5,8d6ac08997bfd848a0a4d9d295e76952"
                    ]
                ],
                "gvcf_tbi": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.g.vcf.gz.tbi:md5,37e2d8f4cca0a21113cede608f54885a"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.vcf.gz:md5,7cba1516f7cf0888765d5ee8caf275f4"
                    ]
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.vcf.gz.tbi:md5,02a78562bc83520a51010a01fb06f217"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,27500aacf5e1da6a8a31905566fcb41f"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T15:30:42.249744"
    }
}